core/src/main/java/org/elasticsearch/search/internal/SearchContext.java

```diff
@@ -67,7 +69,7 @@
 import java.util.Map;
 import java.util.concurrent.atomic.AtomicBoolean;
 
-public abstract class SearchContext implements Releasable, HasContextAndHeaders {
+public abstract class SearchContext extends DelegatingHasContextAndHeaders implements Releasable {
```

Comment:
with this change in place, can you try to remove the `HasContextAndHeaders` delegation methods from `PercolateContext` and `FilteredSearchContext`?